### PR TITLE
Version check failure and socket free fix

### PR DESCRIPTION
**Pull request by pkalever**

### CODE REVIEW

File: daemon/gluster-blockd.c [hunk 1]
```diff
@@ -61,7 +61,7 @@ glusterBlockCliThreadProc (void *vargp)
 {
   register SVCXPRT *transp = NULL;
   struct sockaddr_un saun = {0, };
-  int sockfd = -1;
+  int sockfd = RPC_ANYSOCK;
```
Comment (lxbsz) [inline daemon/gluster-blockd.c]: Should we also change the code like:

   if (sockfd != -1) {  ## -1 --> RPC_ANYSOCK
     close(sockfd);	   
   }
??

**Comment by lxbsz** [comment]
@pkalever 

Tested this PR, works as expected.

One question:

For the Cap checking, currently the code only checks whether the specify option is exist or not. If exist the check will success, or fail. 
Is there the use case that all the options are exist but will set some of them to false ?
And for this case, the current code won't work.

Thanks,

**Comment by pkalever** [comment]
> Tested this PR, works as expected.
> 
Thanks @lxbsz 

> One question:
> 
> For the Cap checking, currently the code only checks whether the specify option is exist or not. If exist the check will success, or fail.
> Is there the use case that all the options are exist but will set some of them to false ?
> And for this case, the current code won't work.
> 
We don't have a case for now, but it should work even with false values. I will test this behavior and 
fix it here, thanks again :-)

**Comment by pkalever** [comment]
@lxbsz it is working as expected for me. This is what I did,

In one of the node changed create_ring_buffer to false in 

\# cat /etc/sysconfig/gluster-block-caps.info
[...]
create_ring_buffer: false
[...]

[root@localhost ~]# gluster-block create sample/block18 ha 3 pralloc full ring-buffer 32 192.168.124.6,192.168.124.163,192.168.124.217 10MiB                                                                       
Version check failed between block servers. (capability 'create_ring_buffer' doesn't exit on 192.168.124.217)                                                                                                      
RESULT:FAIL

Have you restarted gluster-blockd after you change something at /etc/sysconfig/gluster-block-caps.info ?
How did you test, can you please provide the reproducer steps ?

Thanks!

**Comment by lxbsz** [comment]
@pkalever 

I changed this in node#1:
# cat /etc/sysconfig/gluster-block-caps.info
[...]
create_ring_buffer: false
[...]

And then restart it:
# systemctl restart gluster-blockd

Then in node#3:
# while [ 1 ]; do gluster-block create dht0/block3 ha 3 prealloc full ring-buffer 32  192.168.195.157,192.168.195.158,192.168.195.159 1G; gluster-block delete dht0/block3; done
SUCCESSFUL ON:   192.168.195.157 192.168.195.158 192.168.195.159
RESULT: SUCCESS
IQN: iqn.2016-12.org.gluster-block:cfacee74-a48e-4bbf-bf96-73b0c7d24adb
PORTAL(S):  192.168.195.157:3260 192.168.195.158:3260 192.168.195.159:3260
RESULT: SUCCESS
SUCCESSFUL ON:   192.168.195.158 192.168.195.159 192.168.195.157
RESULT: SUCCESS

But this sometimes works for me and only encounter once. A little odd.

**Comment by lxbsz** [comment]
@pkalever 

There is one crash when we comment any of the option then restart the gluster-blockd service.

```
Jul 17 05:23:36 rhel1 gluster-blockd[5370]: Parameter auto_save_on_exit is now 'false'.
Jul 17 05:23:36 rhel1 gluster-blockd[5370]: Parameter logfile is now '/var/log/gluster-block/gluster-block-configshell.log'.
Jul 17 05:23:36 rhel1 gluster-blockd[5370]: Parameter loglevel_file is now 'info'.
Jul 17 05:23:36 rhel1 gluster-blockd[5370]: Parameter auto_enable_tpgt is now 'false'.
Jul 17 05:23:36 rhel1 gluster-blockd[5370]: Parameter auto_add_default_portal is now 'false'.
Jul 17 05:23:36 rhel1 kernel: gluster-blockd[5370]: segfault at c ip 00007f655e7f6ea9 sp 00007fff65bd9110 error 4 in libc-2.17.so[7f655e7aa000+1c3000]
Jul 17 05:23:36 rhel1 systemd[1]: gluster-blockd.service: main process exited, code=killed, status=11/SEGV
```

[0001-cap-fix-crash.txt](https://github.com/gluster/gluster-block/files/2201057/0001-cap-fix-crash.txt)

Thanks,

**Comment by lxbsz** [comment]
@pkalever 
After this fix, here it works now.

Tested-by: Xiubo Li <xiubli@redhat.com>

**Comment by pkalever** [comment]
@lxbsz thanks. As discussed I will pick only line:
-    LOG("mgmt", GB_LOG_ERROR, "gbSetCapabilties: GB_CAP_MAX != %s", count);
+    LOG("mgmt", GB_LOG_ERROR, "gbSetCapabilties: GB_CAP_MAX != %d", count);

From your patch, to fix the crash.

File: daemon/gluster-blockd.c (hunk 1 above)
Comment (pkalever) [reply daemon/gluster-blockd.c]: Done

**Comment by pkalever** [comment]
@lxbsz I have updated the patch with access specifier suggest. Thanks!

File: utils/capabilities.c [hunk 2]
```diff
@@ -93,17 +100,17 @@ gbSetCapabilties(blockResponse **c)
...
     goto out;
   }
 
-  reply->xdata.xdata_len = GB_CAP_MAX * sizeof(gbCapObj);
-  reply->xdata.xdata_val = (char *) caps;
+  globalCapabilities = caps;
 
-  ret = 0;
  out:
   GB_FREE(line);
```
[the remark this answers is not in the document]
Comment (pkalever) [reply utils/capabilities.c]: Done. Thanks!

File: utils/capabilities.c [hunk 3]
```diff
@@ -32,27 +35,29 @@ gbCapabilitiesEnumParse(const char *cap)
...
 
   if (GB_ALLOC_N(caps, GB_CAP_MAX) < 0) {
-    fclose(fp);
-    return -1;
+    LOG("mgmt", GB_LOG_ERROR,
+        "gbSetCapabilties: calloc() failed (%s)", strerror(errno));
+    goto out;
   }
 
   while ((getline(&line, &len, fp)) != -1) {
```
[the remark this answers is not in the document]
Comment (pkalever) [reply utils/capabilities.c]: I cannot think of why to add the count limiting by GB_CAP_MAX. Any strong reason for this ?

[the remark this answers is not in the document]
Comment (pranithk) [reply utils/capabilities.c]: When the file we are reading from erroneously has more than GB_CAP_MAX number of capabilities we should defend for that is what I am thinking.

[the remark this answers is not in the document]
Comment (pkalever) [reply utils/capabilities.c]: This is a read-only file. Hence might not be mandatory ?
Probably can be improved at a separate patch if needed.

[the remark this answers is not in the document]
Comment (pranithk) [reply utils/capabilities.c]: @pkalever It catches human error when we add more than GB_CAP_MAX number of capabilities to the file. If you want to send a separate patch for this, that is fine. But definitely we need to handle this.

[the remark this answers is not in the document]
Comment (pkalever) [reply utils/capabilities.c]: Will raise a new issue for this today.

[the remark this answers is not in the document]
Comment (pranithk) [reply utils/capabilities.c]: https://github.com/gluster/gluster-block/issues/103

File: utils/capabilities.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (pranithk) [reply utils/capabilities.c]: This should be done only in failure cases right?

**Pull request merged**